### PR TITLE
Add support for TUN/TAP interfaces

### PR DESCRIPTION
**Pull request by bengal**
Add support for TUN/TAP devices:

```
$ npc tun1
{
  "name": "tun1",
  "iface_type": "Tun",
  "state": "Down",
  "mtu": 1500,
  "flags": [
    "Broadcast",
    "Multicast"
  ],
  "mac_address": "2A:E2:2E:E4:29:A5",
  "tun": {
    "mode": "Tap",
    "owner": 1001,
    "group": 0,
    "pi": false,
    "vnet_hdr": true,
    "multi_queue": true,
    "persist": true,
    "num_queues": 0,
    "num_disabled_queues": 0
  }
}
```

### CODE REVIEW

File: src/lib/ifaces/iface.rs [hunk 1]
```diff
@@ -227,6 +233,11 @@ pub(crate) fn parse_nl_msg_to_iface(nl_msg: &LinkMessage) -> Option<Iface> {
                         IfaceType::Bridge => {
                             iface_state.bridge = get_bridge_info(&d)
                         }
+                        IfaceType::Tun => {
+                            if let Ok(info) = get_tun_info(&d) {
```
Comment (cathay4t) [inline src/lib/ifaces/iface.rs]: You are ignoring the failure found by `get_bridge_info()`. Please at least use `eprintln!()` to show the errors.

File: src/lib/ifaces/tun.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                IFLA_TUN_OWNER => {
+                    tun_info.owner = Some(parse_as_u32(nla.value()));
+                }
+                IFLA_TUN_GROUP => {
+                    tun_info.group = Some(parse_as_u32(nla.value()));
+                }
+                IFLA_TUN_TYPE => {
+                    match nla.value()[0] {
+                        IFF_TUN => { tun_info.mode = TunMode::Tun; }
+                        IFF_TAP => { tun_info.mode = TunMode::Tap; }
```
Comment (cathay4t) [inline src/lib/ifaces/tun.rs]: The rust way to do it is `tun_info.mode = parse_as_u8(nla.value()).into()`

```rust
impl From<u8> for TunMode {
    fn from(d: u8) -> Self {
    }
}
```

File: tools/test_env [hunk 3]
```diff
@@ -73,6 +73,9 @@ elif [ "CHK$1" == "CHKsriov" ];then
     sudo ip link set eth1 vf 0 vlan 100 mac 36:f7:09:ef:95:f0
     sudo ip link set eth1 vf 1 mac 36:f7:09:ef:95:f1
     sudo ip link set eth1 up
+elif [ "CHK$1" == "CHKtun" ];then
+    sudo ip tuntap add name tun1 mode tap \
```
Comment (cathay4t) [inline tools/test_env:77]: You didn't add `tun1` into `test_env rm` clean up section.

File: src/lib/ifaces/tun.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                }
+                IFLA_TUN_VNET_HDR => {
+                    tun_info.vnet_hdr = parse_as_u8(nla.value()) > 0
+                }
+                IFLA_TUN_PERSIST => {
+                    tun_info.persist = parse_as_u8(nla.value()) > 0
+                }
+                IFLA_TUN_MULTI_QUEUE => {
+                    tun_info.multi_queue = parse_as_u8(nla.value()) > 0
+                }
```
Comment (cathay4t) [inline src/lib/ifaces/tun.rs]: They are not ended with `;` which may be cause rust compiler or clippy warning. Please check with `rust clippy` results.

**Review by cathay4t — CHANGES_REQUESTED** [review]
Thanks for the great work. Just some trivial to fix.

File: src/lib/ifaces/iface.rs (hunk 1 above)
Comment (bengal) [reply src/lib/ifaces/iface.rs]: How is it now?

File: src/lib/ifaces/tun.rs (hunk 2 above)
Comment (bengal) [reply src/lib/ifaces/tun.rs]: Changed, thanks for the tip.

File: tools/test_env (hunk 3 above)
Comment (bengal) [reply tools/test_env:77]: Fixed.

File: src/lib/ifaces/tun.rs (hunk 4 above)
Comment (bengal) [reply src/lib/ifaces/tun.rs]: Fixed.

**Review by cathay4t — APPROVED** [review]
LGTM.

**Pull request merged**